server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/http/SessionSupplier.scala

```diff
@@ -0,0 +1,108 @@
+/** **************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ * *
+ * http://www.apache.org/licenses/LICENSE-2.0                 *
+ * *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ * ***************************************************************/
+
+package org.apache.james.jmap.http
+
+import java.net.URL
+
+import com.google.common.annotations.VisibleForTesting
+import eu.timepit.refined.auto._
+import eu.timepit.refined.refineV
+import org.apache.james.core.Username
+import org.apache.james.jmap.http.SessionSupplier.{CORE_CAPABILITY, MAIL_CAPABILITY}
+import org.apache.james.jmap.model.CapabilityIdentifier.CapabilityIdentifier
+import org.apache.james.jmap.model.Id.Id
+import org.apache.james.jmap.model._
+import reactor.core.publisher.Mono
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+object SessionSupplier {
+  private val CORE_CAPABILITY = CoreCapability(
+    properties = CoreCapabilityProperties(
+      MaxSizeUpload(10_000_000L),
+      MaxConcurrentUpload(4L),
+      MaxSizeRequest(10_000_000L),
+      MaxConcurrentRequests(4L),
+      MaxCallsInRequest(16L),
+      MaxObjectsInGet(500L),
+      MaxObjectsInSet(500L),
+      collationAlgorithms = List("i;unicode-casemap")))
+
+  private val MAIL_CAPABILITY = MailCapability(
+    properties = MailCapabilityProperties(
+      MaxMailboxesPerEmail(Some(10_000_000L)),
+      MaxMailboxDepth(None),
+      MaxSizeMailboxName(200L),
+      MaxSizeAttachmentsPerEmail(20_000_000L),
+      emailQuerySortOptions = List("receivedAt", "cc", "from", "to", "subject", "size", "sentAt", "hasKeyword", "uid", "Id"),
+      MayCreateTopLevelMailbox(true)
+    ))
+}
+
+class SessionSupplier {
+  def generate(username: Username): SMono[Session] =
+    SMono.fromPublisher(
```

Comment:
use SMono.zip directly

Reply:
`SMono.zip` is hard to use, it requires an iterable as the first parameter and then the combinator using that iterable in the second parameter. Not easy as Mono java.

Can you check?

Reply:
Yes, it needs some ugly things to work.
If you can live with the fact that if a Mono fail the Mono will wait for the other before completing , you can do this :
```
  def generate(username: Username): SMono[Session] = {
      SMono.zipDelayError(
        accounts(username),
        primaryAccounts(username))
      .map(tuple => generate(username, tuple._1, tuple._2))
  }
```

otherwise your solution is ok to me

Reply:
Oh no I want to fail fast! but thanks for the suggestion!